File: perturb.f90
subroutine wave_prep(N, M, lambda_k, A, phi)

 ! =====================================================
 !  Accepts number of harmonics, N, and max wavelength, lambda_k,
 !  then returns randomized amplitude, A, and phase shift, phi
 ! =====================================================

implicit none

integer, intent(in)                                 :: lambda_k, N, M
real(8), intent(out), dimension(-N:N, -M:M)         :: phi
real(8), intent(out), dimension(-N:N, -M:M)         :: A
real(8)                                             :: random_normal, randn
integer                                             :: i, j

do i =  -N, N
    do j = -M, M
        if (((i**2 + j**2) .gt. 0) .and. ((i**2 + j**2) .lt. N**2 + 1)) then
            randn = random_normal()
            A(i, j) = (lambda_k - randn) * ((real(i)**2 + real(j)**2)/(N**2))
        else
            A(i, j) = 0
        end if
        randn = random_normal()
        phi(i, j) = 10 * randn
    end do
end do

end subroutine



subroutine axi_fourier_series(pts, phi, A, lambda_k, sample_points, N, M, pts_perturbed)

 ! =====================================================
 !  Accepts x y z points in the pts array, with other
 !  parameters for the fourier series perturbation process
 ! =====================================================

implicit none

integer, intent(in)                                       :: sample_points, N, M
real(8), intent(in)                                       :: lambda_k
real(8), intent(in), dimension(-N:N, -M:M)                :: phi, A
real(8), intent(in), dimension(0:sample_points-1,3)       :: pts
real(8), intent(out), dimension(0:sample_points-1,3)      :: pts_perturbed
real(8), dimension(-N:N, -N:N)                            :: N_mat, M_mat, B
real(8), dimension(1:2)                                   :: bounds
real(8), dimension(0:sample_points-1)                     :: x, y, z, r, th, double_sum
real(8)                                                   :: pi = 4.0 * ATAN(1.0)
real(8)                                                   :: max_height
real(8)                                                   :: half_height
real(8)                                                   :: inv_maxr
real(8)                                                   :: inv_lambda
real(8)                                                   :: trans_point
real(8)                                                   :: min_radius
real(8)                                                   :: scale_height
real(8)                                                   :: inv_scale_height
integer                                                   :: i, j

do i = 0, sample_points-1
  x(i) = pts(i,1) * 1000     ! axial coordinates in mils
  y(i) = pts(i,2) * 1000     ! radial coordinates in mils
  z(i) = pts(i,3) * 1000     ! radial coordinates in mils
end do

! Set axial boundaries (stagnation + 10 mils to [end - X] mils)
bounds = (/minval(x(:))+10, maxval(x(:))-250/)

! Map radial coordinates into cylindrical coordinate system
do i = 0, sample_points-1
  th(i) = atan2(z(i),y(i))
  r(i) = sqrt(y(i)**2 + z(i)**2)
end do

! Create N_mat and M_mat to avoid a triple do loop
do i = -N, N
  N_mat(i,:) = i
end do

do j = -M, M
  M_mat(:,j) = j
end do

! Set limits and constants for personalization and speed
! - max_height limits perturbation amplitudes to this quantity
! - trans_point is a location upstream of internal geometry that should not be perturbed (units are mils)
! - min_radius is a location outside of internal geometry that should not be perturbed
! - inv_maxr and inv_lambda are used for speed due to cost of division
! - scale_height is a scaling factor to reduce the double sum
! - inv_scale_height is a dividor to reduce the double sum
max_height          = 1.2 * lambda_k 
half_height         = 0.5 * lambda_k
trans_point         = 5750
min_radius          = 650
inv_maxr            = 1 / maxval(r(:))
inv_lambda          = 1 / (lambda_k * N)
scale_height        = 2*N
inv_scale_height    = 1 / scale_height

! Perturbation Loop
! - Perturb all sampled points in the radial direction between bounds
! - Give flat tops to any peaks that surpassed max_height due to frequency doubling
! - Else if statement avoids perturb internal geometry
! - r(i) * inv_maxr scales perturbations linearly from the min to max radius
do i = 0, sample_points-1
    if ((x(i).gt.bounds(1)).and.(x(i).le.(trans_point))) then
        B = cos((2 * pi * N_mat * th(i) * inv_lambda) + (2 * pi * M_mat * x(i) * inv_lambda) + phi)
        double_sum(i) = sum(A*B) * inv_scale_height
!        if (abs(double_sum(i)) .gt. max_height) then
!           double_sum(i) = max_height
!        end if
        r(i) = r(i) + abs(double_sum(i)) * r(i) * inv_maxr
    else if ((x(i).gt.(trans_point)).and.(x(i).le.(bounds(2))).and.(r(i).gt.(min_radius))) then
        B = cos((2 * pi * N_mat * th(i) * inv_lambda) + (2 * pi * M_mat * x(i) * inv_lambda) + phi)
        double_sum(i) = sum(A*B) * inv_scale_height
!        if (abs(double_sum(i)) .gt. max_height) then
!            double_sum(i) = max_height
!        end if
        r(i) = r(i) + abs(double_sum(i)) * r(i) * inv_maxr
    end if
end do

! Convert coordinates back to cartesian
x(:) = x(:) / 1000              ! axial coordinates, mils to inches
y(:) = r(:) * cos(th(:)) / 1000 ! radial coordinates, mils to inches
z(:) = r(:) * sin(th(:)) / 1000 ! radial coordinates, mils to inches

! Gather perturbed x y z coordinates
do i = 0, sample_points-1
  pts_perturbed(i,1) = x(i)
  pts_perturbed(i,2) = y(i)
  pts_perturbed(i,3) = z(i)
end do

end subroutine



subroutine cart_fourier_series(pts, phi, A, lambda_k, sample_points, N, M, x, y, z)

 ! =====================================================
 !  Accepts x and y points in the pts array, with other
 !  parameters for the fourier series perturbation process
 ! =====================================================

implicit none

integer, intent(in)                                                             :: sample_points, N, M
real(8), intent(in)                                                             :: lambda_k
real(8), intent(in), dimension(-N:N, -M:M)                                      :: phi, A
real(8), intent(in), dimension(0:(sample_points)-1,2)                           :: pts
real(8), intent(out), dimension(0:(sample_points)-1)                            :: x, y, z
real(8), dimension(0:sample_points-1)                                           :: double_sum
real(8), dimension(-N:N, -N:N)                                                  :: N_mat, M_mat, B
real(8)                                                                         :: pi = 4.0 * ATAN(1.0)
real(8)                                                                         :: inv_lambda
real(8)                                                                         :: scale_height
real(8)                                                                         :: inv_scale_height
integer                                                                         :: i, j

do i = 0, sample_points-1
  x(i) = pts(i,1)      ! x coordinates in mils
  y(i) = pts(i,2)      ! y coordinates in mils
end do

! Create N_mat and M_mat to avoid a triple do loop
do i = -N, N
  N_mat(i,:) = i
end do

do j = -M, M
  M_mat(:,j) = j
end do

inv_lambda          = 1 / (lambda_k * N)
scale_height        = 2*N
inv_scale_height    = 1 / scale_height

! Perturbation Loop
do i = 0, sample_points-1
        B = cos((2 * pi * N_mat * x(i) * inv_lambda) + (2 * pi * M_mat * y(i) * inv_lambda) + phi)
        double_sum(i) = sum(A*B) * inv_scale_height
        z(i) = abs(double_sum(i))
end do

! Convert coordinates back to inches
x(:) = x(:) / 1000 
y(:) = y(:) / 1000 
z(:) = z(:) / 1000 

end subroutine



real(8) function random_normal()

 ! =====================================================
 !  Calculates pseudo-random normal (gaussian) numbers
 ! =====================================================

implicit none

  real(8)                   :: r1
  real(8)                   :: r2
  real(8)                   :: x
  real(8)                   :: pi = 4.0 * ATAN(1.0)

  r1 = rand()
  r2 = rand()
  x = sqrt ( - 2.0E+00 * log ( r1 ) ) * cos ( 2.0E+00 * pi * r2 )
  random_normal = x

return
end function random_normal

File: perturb.py
import numpy as np

def wave_prep(N, M, lambda_k):
    """
    Accepts number of harmonics, N and M, and min wavelength, lambda_k,
    then returns randomized amplitude, A, and phase shift, phi
    """
    
    A = np.zeros([2*N+1,2*M+1])
    phi = np.zeros([2*N+1,2*M+1])
    for i, n in enumerate(np.arange(-N, N+1)):
        for j, m in enumerate(np.arange(-M, M+1)):
            if ((n**2 + m**2) < (N**2 + 1)):
                A[i][j] = lambda_k / 2 * (n**2 + m**2)/(N**2)
            else:
                A[i][j] = 0
            phi[i][j] = lambda_k * np.random.randn() / 4
    
    return A, phi



def axi_fourier_series(pts, phi, A, lambda_k, sample_points, N, M):
    """
    Accepts x y z points in the pts array, with other
    parameters for the fourier series perturbation process
    """
    
    # convert coordinates from inches to mils
    x = pts[:,0] * 1000
    y = pts[:,1] * 1000
    z = pts[:,2] * 1000
    
    # Set axial boundaries for linear tapering
    bounds = [50, 550, 6250, 6750]
    
    # Map cartesian coordinates into cylindrical coordinate system
    th = np.zeros(sample_points)
    r  = np.zeros(sample_points)
    for i in np.arange(sample_points):
      th[i] = np.arctan2(z[i], y[i])
      r[i]  = np.sqrt(y[i]**2 + z[i]**2)
    
    # Create N_mat and M_mat to avoid a triple do loop
    N_mat = np.zeros([2*N+1, 2*M+1])
    for i, n in enumerate(np.arange(-N,N+1)):
        N_mat[i,:] = n
    
    M_mat = np.zeros([2*N+1, 2*M+1])
    for j, m in enumerate(np.arange(-M,M+1)):
        M_mat[:,j] = m
    
    # Set limits and constants for personalization and speed
    # max_height limits perturbation amplitudes to this quantity
    # trans_point is a location upstream of internal geometry that should not be perturbed (units are mils)
    # min_radius is a location outside of internal geometry that should not be perturbed
    # inv_maxr and inv_lambda are used for speed due to cost of division
    # scale_height is a scaling factor to reduce the double sum
    
    max_height    = 1.25 * lambda_k 
    trans_point   = 5000
    min_radius    = 650
    inv_maxr      = 1 / np.max(r[:])
    inv_maxx      = 1 / bounds[1]
    inv_lambda_th = 1 / (lambda_k * N)
    inv_lambda_x  = 1 / (lambda_k * M)
    
    # Perturbation Loop
    # Perturb all sampled points in the radial direction between bounds
    # Else if statement avoids perturbing internal geometry
    # r(i) * inv_maxr scales perturbations linearly from the min to max radius
    B = np.zeros([2*N+1,2*M+1])
    double_sum = np.zeros(sample_points)
    for i in np.arange(sample_points):
        
        if  bounds[0] < x[i] and x[i] <= bounds[1]:
            B = np.cos((2 * np.pi * N_mat * th[i] * r[i] * inv_lambda_th) + (2 * np.pi * M_mat * x[i] * inv_lambda_x) + phi)
            double_sum[i] = np.sum(A*B) / (2*N+1)
            if (double_sum[i] > max_height):
                double_sum[i] = max_height
            r[i] = r[i] + (double_sum[i] + lambda_k/2) * (x[i] - bounds[0]) / (bounds[1] - bounds[0])
        
        elif bounds[1] < x[i] and x[i] <= trans_point:
            B = np.cos((2 * np.pi * N_mat * th[i] * r[i] * inv_lambda_th) + (2 * np.pi * M_mat * x[i] * inv_lambda_x) + phi)
            double_sum[i] = np.sum(A*B) / (2*N+1)
            if (double_sum[i] > max_height):
                double_sum[i] = max_height
            r[i] = r[i] + (double_sum[i] + lambda_k/2)
        
        elif trans_point < x[i] and x[i] <= bounds[2] and r[i] > min_radius:
            B = np.cos((2 * np.pi * N_mat * th[i] * r[i] * inv_lambda_th) + (2 * np.pi * M_mat * x[i] * inv_lambda_x) + phi)
            double_sum[i] = np.sum(A*B) / (2*N+1)
            if (double_sum[i] > max_height):
                double_sum[i] = max_height
            r[i] = r[i] + (double_sum[i] + lambda_k/2)
        
        elif bounds[2] < x[i] and x[i] <= bounds[3] and r[i] > min_radius:
            B = np.cos((2 * np.pi * N_mat * th[i] * r[i] * inv_lambda_th) + (2 * np.pi * M_mat * x[i] * inv_lambda_x) + phi)
            double_sum[i] = np.sum(A*B) / (2*N+1)
            if (double_sum[i] > max_height):
                double_sum[i] = max_height
            r[i] = r[i] + (double_sum[i] + lambda_k/2) * (bounds[3] - x[i]) / (bounds[3] - bounds[2])
    
    # Convert coordinates back to cartesian
    x[:] = x[:]                 / 1000 
    y[:] = r[:] * np.cos(th[:]) / 1000 
    z[:] = r[:] * np.sin(th[:]) / 1000 
    
    pts_perturbed = np.zeros([sample_points,3])
    pts_perturbed[:,0] = x[:]
    pts_perturbed[:,1] = y[:]
    pts_perturbed[:,2] = z[:]
    
    return pts_perturbed
    
    
    
def cart_fourier_series(pts, phi, A, lambda_k, sample_points, N, M):
    """
    Accepts the pts array, with other parameters for the fourier series 
    perturbation process and returns x y z for the perturbed solution
    """
    
    x = pts[:,0]
    y = pts[:,1]
    
    # Create N_mat and M_mat to avoid a triple do loop
    N_mat = np.zeros([2*N+1, 2*M+1])
    for i, n in enumerate(np.arange(-N,N+1)):
        N_mat[i,:] = n
    
    M_mat = np.zeros([2*N+1, 2*M+1])
    for j, m in enumerate(np.arange(-M,M+1)):
        M_mat[:,j] = m
    
    inv_lambda_x   = 1 / (lambda_k * N)
    inv_lambda_y   = 1 / (lambda_k * M)

    # Perturbation Loop
    B = np.zeros([2*N+1,2*M+1])
    z = np.zeros(sample_points)
    for i in np.arange(sample_points):
        B = np.cos((2 * np.pi * N_mat * x[i] * inv_lambda_x) + (2 * np.pi * M_mat * y[i] * inv_lambda_y) + phi)
        z[i] = np.sum(A*B) / (2*N+1)
    
    # Convert coordinates back to inches
    x[:] = x[:] / 1000
    y[:] = y[:] / 1000
    z[:] = z[:] / 1000

    return x, y, z